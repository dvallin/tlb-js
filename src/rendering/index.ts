import { Vector2D } from "@/geometry/Vector2D"
import LazyJS from "lazy.js"

export type RenderIterator = () => Vector2D | undefined

export function toStream(iter: RenderIterator): LazyJS.Sequence<Vector2D> {
    return LazyJS.generate(iter)
        .takeWhile((p: Vector2D | undefined) => p !== undefined)
        .map(p => p as Vector2D)
}

export function toArray(iter: RenderIterator): Vector2D[] {
    const array = []
    while (true) {
        const p: Vector2D | undefined = iter()
        if (p === undefined) {
            return array
        }
        array.push(p)
    }
}

export function foreach(iter: RenderIterator, callback: (p: Vector2D) => void): void {
    while (true) {
        const p: Vector2D | undefined = iter()
        if (p === undefined) {
            return
        }
        callback(p)
    }
}
