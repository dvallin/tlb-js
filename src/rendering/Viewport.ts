import { DEFAULT_HEIGHT, DEFAULT_WIDTH } from "@/Game"
import { Position } from "@/geometry/Position"
import { GameSystem, RenderLayer } from "@/systems/GameSystem"
import { World, Boxed } from "mogwai-ecs/lib"
import { Display } from "rot-js"

import { Rectangle } from "@/geometry/Rectangle"
import { Size } from "@/geometry/Size"
import { Input } from "@/systems/Input"
import { MapSystem } from "@/map/Map"
import { MenuSystem, MenuItems } from "@/menu/Menu"

export class Viewport {
    public constructor(
        public offset: Position,
        public rectangle: Rectangle
    ) { }

    public get topLeft(): Position {
        return new Position(this.rectangle.left, this.rectangle.top).subtract(this.offset)
    }
}

export class ViewportSystem implements GameSystem {

    public static NAME: string = "viewport"

    public renderLayer: RenderLayer = RenderLayer.None

    public menuViewport: Viewport = new Viewport(
        new Position(0, 0),
        Rectangle.from(
            new Position(0, 0),
            new Size(DEFAULT_WIDTH, 1)
        )
    )

    public mapViewport: Viewport = new Viewport(
        new Position(0, 1),
        Rectangle.from(
            new Position(0, 0),
            new Size(DEFAULT_WIDTH, DEFAULT_HEIGHT - 1)
        )
    )

    public register(world: World): void {
        world.registerSystem(ViewportSystem.NAME, this)
    }

    public build({ }: World): void {
        //
    }

    public execute(world: World): void {
        const input: Input | undefined = world.systems.get(Input.NAME) as Input | undefined
        const menu: MenuSystem | undefined = world.systems.get(MenuSystem.NAME) as MenuSystem | undefined
        if (input !== undefined && menu !== undefined) {
            switch (menu.activeMenuItem) {
                case MenuItems.Map:
                    this.moveMapViewport(input)
                    break
                case MenuItems.Player:
                    this.focusOnPlayer(world)
                    break
            }
        }

        const map: MapSystem | undefined = world.systems.get(MapSystem.NAME) as MapSystem | undefined
        if (map !== undefined) {
            this.mapViewport.rectangle = this.mapViewport.rectangle.clamp(map.boundary)
        }
    }

    public render({ }: World, { }: Display): void {
        //
    }

    private moveMapViewport(input: Input): void {
        const delta = input.movementDelta()
        this.mapViewport.rectangle = this.mapViewport.rectangle.add(delta.round())

        if (input.mouse.left || input.mouse.right) {
            const mouseDrag = new Position(
                (input.mouse.x - input.mouse.clickX!) * 0.2,
                (input.mouse.y - input.mouse.clickY!) * 0.2,
            )
            this.mapViewport.rectangle = this.mapViewport.rectangle.add(mouseDrag.round())
        }
    }

    private focusOnPlayer(world: World): void {
        const player: { position: Boxed<Position> } | undefined = world
            .fetch()
            .on(t => t.hasLabel("player").hasLabel("active"))
            .withComponents("position")
            .first()
        if (player !== undefined) {
            this.mapViewport.rectangle = this.mapViewport.rectangle.focus(player.position.value)
        }
    }
}
