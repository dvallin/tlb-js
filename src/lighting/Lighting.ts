import { DEFAULT_HEIGHT, DEFAULT_WIDTH } from "@/Game"
import { Position, Domain } from "@/geometry/Position"
import { GameSystem, RenderLayer } from "@/systems/GameSystem"
import { World, Boxed } from "mogwai-ecs/lib"
import { Display, FOV, Lighting } from "rot-js"

import { Tile } from "@/map/Tile"
import { foreach } from "@/rendering"
import { Rectangle } from "@/geometry/Rectangle"
import { rasterize as rasterizeRectangle } from "@/rendering/rectangle"
import { Size } from "@/geometry/Size"
import { gray } from "@/rendering/palettes"
import { Color } from "@/rendering/Color"
import { MenuSystem, MenuItems } from "@/menu/Menu"
import { MapSystem } from "@/map/Map"
import { Drawable } from "@/rendering/Drawable"
import { Vector2D } from "@/geometry/Vector2D"

export class LightingSystem implements GameSystem {

    public static NAME: string = "lighting"

    public readonly boundary: Size = new Size(2 * DEFAULT_WIDTH, 2 * DEFAULT_HEIGHT)
    public renderLayer: RenderLayer = RenderLayer.Layer3

    private lightBlocking: Set<string> = new Set()
    private entityMap: Map<string, Drawable[]> = new Map()
    private visible: Set<string> = new Set()
    private discovered: Set<string> = new Set()

    private ambientLight: Color = gray[3]
    private lightingEnabled: boolean = true

    public register(world: World): void {
        world.registerSystem(LightingSystem.NAME, this)
        world.registerComponent("light")
        world.registerComponent("lightBlocking")
        world.registerRelation("contains")
        world.registerRelation("connected")
    }

    public build(world: World): void {
        const map: MapSystem | undefined = world.systems.get(MapSystem.NAME) as MapSystem | undefined
        if (map !== undefined) {
            if (!map.built) {
                map.build(world)
            }

            const lights: { entity: number, position: Boxed<Position> }[] = world
                .fetch()
                .on(t => t.hasLabel("light")).withComponents("position")
                .collect()
            const domains = new Map<Domain, { entity: number, position: Boxed<Position> }[]>()
            lights.forEach(light => {
                const domain = domains.get(light.position.value.domain) || []
                domain.push(light)
                domains.set(light.position.value.domain, domain)
            })
            domains.forEach((value, key) => {
                this.updateLighting(world, key, value, [Rectangle.from(new Vector2D(0, 0), this.boundary)])
            })
        }
    }

    public execute(world: World): void {
        const menu: MenuSystem | undefined = world.systems.get(MenuSystem.NAME) as MenuSystem | undefined
        const map: MapSystem | undefined = world.systems.get(MapSystem.NAME) as MapSystem | undefined
        this.lightingEnabled = true
        if (menu !== undefined && menu.activeMenuItem === MenuItems.Map) {
            this.lightingEnabled = false
        }
        if (this.lightingEnabled) {
            const playerPosition: Boxed<Position> = world.fetch()
                .on(t => t.hasLabel("player").hasLabel("active"))
                .withComponents("position")
                .first()
                .position
            const domain = playerPosition.value.domain

            const playerTile = map!.getTile(playerPosition.value)
            if (playerTile !== undefined && playerTile.room !== undefined) {
                this.visible.clear()
                const roomIds: Set<number> = new Set()
                const fov = new FOV.RecursiveShadowcasting(
                    (x, y) => this.isLightPassing(map!, new Position(domain, new Vector2D(x, y))),
                    { topology: 8 }
                )
                fov.compute(playerPosition.value.x, playerPosition.value.y, 20, (x, y) => {
                    const index = map!.getIndex(new Position(domain, new Vector2D(x, y)))
                    if (index !== undefined) {
                        this.discovered.add(index)
                        this.visible.add(index)
                        const room = map!.getTileByIndex(index)!.room
                        if (room !== undefined) {
                            roomIds.add(room)
                        }
                    }
                })

                const lights: { entity: number, position: Boxed<Position> }[] = world
                    .fetch(...Array.from(roomIds))
                    .on(t => t.out("contains").hasLabel("light"))
                    .withComponents("position")
                    .collect()
                const dirtyTiles: Rectangle[] = world
                    .fetch(...Array.from(roomIds))
                    .withComponents("room")
                    .stream()
                    .map(p => p.room.shape.grow(1))
                    .toArray()
                this.updateLighting(world, domain, lights, dirtyTiles)

            }
        }
    }

    public render({ }: World, { }: Display): void {
        //
    }

    public afterRender({ }: World): void {
        //
    }

    public buildLight(world: World, position: Position): number {
        return world.entity()
            .with("light")
            .with("position", new Boxed(position))
            .close()
    }

    public isVisible(index: string): boolean {
        return this.visible.has(index)
    }

    public isDiscovered(index: string): boolean {
        return this.discovered.has(index)
    }

    public setDrawable(drawable: Drawable): void {
        drawable.computeColor(this.ambientLight)
    }

    public isLightPassing(map: MapSystem, position: Position): boolean {
        const index: string | undefined = map.getIndex(position)
        if (index === undefined) {
            return false
        }

        const tile: Tile | undefined = map.getTileByIndex(index)
        if (tile === undefined || tile.lightBlocking) {
            return false
        }

        if (this.lightBlocking.has(index)) {
            return false
        }

        return true
    }

    public getColor(drawable: Drawable): Color {
        if (!this.lightingEnabled) {
            return drawable.diffuse
        }
        return drawable.color
    }

    private updateLighting(world: World, domain: Domain,
        lights: { entity: number, position: Boxed<Position> }[],
        dirtyAreas: Rectangle[]
    ): void {
        const map: MapSystem | undefined = world.systems.get(MapSystem.NAME) as MapSystem | undefined
        if (map === undefined) {
            return
        }

        this.lightBlocking.clear()
        world.fetch()
            .on(t => t.hasLabel("lightBlocking").hasLabel("position"))
            .withComponents("position")
            .stream()
            .each((v: { position: Boxed<Position> }) => {
                const idx = map!.getIndex(v.position.value)
                if (idx !== undefined) {
                    this.lightBlocking.add(idx)
                }
            })

        this.entityMap.clear()
        world.fetch()
            .on(t => t.hasLabel("drawable").hasLabel("position"))
            .withComponents("position", "drawable")
            .stream()
            .each((v: { position: Boxed<Position>, drawable: Drawable }) => {
                const idx = map!.getIndex(v.position.value)
                v.drawable.clearLight()
                if (idx !== undefined) {
                    if (this.entityMap.has(idx)) {
                        this.entityMap.get(idx)!.push(v.drawable)
                    } else {
                        this.entityMap.set(idx, [v.drawable])
                    }
                }
            })

        const dirtyTiles: Set<string> = new Set<string>()
        for (const area of dirtyAreas) {
            foreach(rasterizeRectangle(area, true), (v) => {
                const p: Position = new Position(domain, v)
                const tile: Tile | undefined = map.getTile(p)
                if (tile !== undefined) {
                    for (const light of lights) {
                        if (tile.hasLight(light.entity)) {
                            tile.removeLight(light.entity)
                            dirtyTiles.add(map.getIndex(p)!)
                        }
                    }
                }
            })
        }

        for (const light of lights) {
            const fov = new FOV.PreciseShadowcasting(
                (x, y) => this.isLightPassing(map, new Position(domain, new Vector2D(x, y))),
                { topology: 8 }
            )
            const lighting = new Lighting(
                (x, y) => this.isLightPassing(map, new Position(domain, new Vector2D(x, y))) ? 1.0 : 0.0,
                { passes: 1 }
            )
            lighting.setLight(light.position.value.x, light.position.value.y, [255, 255, 255])
            lighting.setFOV(fov)
            lighting.compute((x: number, y: number, c: [number, number, number]) => {
                const p = new Position(domain, new Vector2D(x, y))
                const index = map.getIndex(p)
                if (index !== undefined) {
                    const tile: Tile | undefined = map.getTileByIndex(index)
                    const color = new Color(c)
                    if (tile !== undefined) {
                        dirtyTiles.add(map.getIndex(p)!)
                        tile.setLight(light.entity, color)
                    }
                    if (this.entityMap.has(index)) {
                        this.entityMap.get(index)!.forEach(drawable =>
                            drawable.setLight(light.entity, color)
                        )
                    }
                }
            })
        }

        dirtyTiles.forEach(index => {
            const tile = map.getTileByIndex(index)
            if (tile !== undefined) {
                tile.computeColor(this.ambientLight)
            }
            if (this.entityMap.has(index)) {
                this.entityMap.get(index)!.forEach(drawable =>
                    drawable.computeColor(this.ambientLight)
                )
            }
        })
    }
}
