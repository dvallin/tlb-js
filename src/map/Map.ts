import { Position, Domain } from "@/geometry/Position"
import { GameSystem, RenderLayer } from "@/systems/GameSystem"
import { MapStorage, World, Boxed, PartitionedStorage, VectorStorage } from "mogwai-ecs/lib"
import { Display } from "rot-js"

import { Tile, wallTile, roomTile, hubTile, corridorTile } from "@/map/Tile"
import { Drawable } from "@/rendering/Drawable"
import { foreach, toArray } from "@/rendering"
import { Rectangle } from "@/geometry/Rectangle"
import { rasterize as rasterizeRectangle } from "@/rendering/rectangle"
import { Size } from "@/geometry/Size"
import { Input } from "@/systems/Input"
import { gridSymbols } from "@/symbols"
import { gray, primary } from "@/rendering/palettes"
import { ViewportSystem, Viewport } from "@/rendering/Viewport"
import { Color } from "@/rendering/Color"
import { TunnelingBuilder } from "@/map/generators/TunnelingBuilder"
import { MenuSystem, MenuItems } from "@/menu/Menu"
import { LightingSystem } from "@/lighting/Lighting"
import { doorTrigger, Trigger } from "@/triggers/TriggerSystem"
import { Vector } from "@/geometry/Vector"
import { Vector2D } from "@/geometry/Vector2D"
import { GameSettings } from "@/Game";

interface DrawableWithData {
    position: Boxed<Position>, description: Boxed<string> | undefined, drawable: Drawable
}

interface Room {
    shape: Rectangle
}

export class SparseMap<T> {
    private map: { [index: string]: T } = {}

    public constructor(
        private boundary: Vector
    ) { }

    public set(position: Position, value: T): boolean {
        const index = this.index(position)
        if (index === undefined) {
            return false
        }
        this.map[index] = value
        return true
    }

    public get(position: Position): T | undefined {
        const index = this.index(position)
        if (index === undefined) {
            return undefined
        }
        return this.getByIndex(index)
    }

    public getByIndex(index: string): T | undefined {
        return this.map[index]
    }

    public index(position: Position): string | undefined {
        if (!this.inside(position)) {
            return undefined
        }
        return position.index()
    }

    public inside(position: Position): boolean {
        return position.inside(this.boundary)
    }
}

export class MapSystem implements GameSystem {

    public static NAME: string = "map"

    public static fromSettings(settings: GameSettings): MapSystem {
        return new MapSystem(settings.map_width, settings.map_height)
    }

    public built: boolean = false

    public readonly boundary: Size
    public renderLayer: RenderLayer

    private map: SparseMap<Tile>
    private drawables: Map<string, Drawable[]>
    private selectedRoom: number | undefined
    private activeDomain: Domain | undefined

    private constructor(width: number, height: number) {
        this.boundary = new Size(width, height)
        this.map = new SparseMap(this.boundary)
        this.renderLayer = RenderLayer.Layer2
        this.drawables = new Map()
    }

    public register(world: World): void {
        world.registerSystem(MapSystem.NAME, this)
        world.registerComponent("drawable", new MapStorage<Drawable>())
        world.registerComponent("blocking")
        world.registerComponent("active", new MapStorage<Tile>())
        world.registerComponent("description", new MapStorage<Boxed<string>>())
        world.registerComponent("trigger", new MapStorage<Trigger>())
        world.registerComponent("position", new PartitionedStorage(
            new VectorStorage<Boxed<Position>>(),
            (value: Position) => this.map.index(value) || ""
        ))

        world.registerComponent("room", new MapStorage<Room>())
        world.registerRelation("contains")
        world.registerRelation("connected")
        world.registerComponent("light")
        world.registerComponent("lightBlocking")
    }

    public build(world: World): void {
        if (!this.built) {
            this.activeDomain = Domain.Tower
            const entrance = new Position(this.activeDomain, new Vector2D(Math.floor(this.boundary.width / 2), 0))
            new TunnelingBuilder(world, this)
                .startAt(entrance)
                .run()
            this.fillWalls(this.activeDomain!)
            this.built = true
        }
    }

    public execute(world: World): void {
        const input: Input | undefined = world.systems.get(Input.NAME) as Input | undefined
        const viewport: ViewportSystem | undefined = world.systems.get(ViewportSystem.NAME) as ViewportSystem | undefined

        const playerPosition: Boxed<Position> = world.fetch()
            .on(t => t.hasLabel("player").hasLabel("active"))
            .withComponents("position")
            .first()
            .position
        this.activeDomain = playerPosition.value.domain

        if (input !== undefined && viewport !== undefined) {
            const mousePosition = new Vector2D(input.mouse.x, input.mouse.y)
            const topLeft = viewport.mapViewport.topLeft
            const tile = this.getTile(new Position(this.activeDomain, mousePosition.add(topLeft)))
            if (tile !== undefined) {
                this.selectedRoom = tile.room
            }
        }
        this.drawables.clear()
        world.fetch()
            .on(t => t.hasLabel("drawable").hasLabel("position"))
            .withComponents("drawable", "position")
            .stream()
            .each((v: { position: Boxed<Position>, drawable: Drawable }) => {
                const index = this.map.index(v.position.value)
                if (index !== undefined) {
                    if (!this.drawables.has(index)) {
                        this.drawables.set(index, [v.drawable])
                    } else {
                        this.drawables.get(index)!.push(v.drawable)
                    }
                }
            })
    }

    public render(world: World, display: Display): void {
        const viewport: ViewportSystem | undefined = world.systems.get(ViewportSystem.NAME) as ViewportSystem | undefined
        const lighting: LightingSystem | undefined = world.systems.get(LightingSystem.NAME) as LightingSystem | undefined
        if (viewport !== undefined && lighting !== undefined) {
            const topLeft = viewport.mapViewport.topLeft
            foreach(rasterizeRectangle(viewport.mapViewport.rectangle, true), v => {
                const position = new Position(this.activeDomain!, v)
                const index: string | undefined = this.map.index(position)
                if (index !== undefined && lighting.isDiscovered(index)) {
                    const tile: Tile | undefined = this.getTileByIndex(index)
                    if (tile !== undefined) {
                        const mapPosition = position.subtract(topLeft)
                        const color = lighting.getColor(tile, index)
                        this.renderTile(display, mapPosition, tile, color)
                    }
                }
            })

            this.drawDrawables(world, display, viewport.mapViewport)
        }
    }

    public afterRender({ }: World): void {
        //
    }

    public setTile(world: World, position: Position, tile: Tile): void {
        const valid = this.map.set(position, tile)
        if (valid) {
            const lighting: LightingSystem | undefined = world.systems.get(LightingSystem.NAME) as LightingSystem | undefined
            if (lighting !== undefined) {
                lighting.setDrawable(tile)
            }
        }
    }

    public getTile(position: Position): Tile | undefined {
        return this.map.get(position)
    }

    public getIndex(position: Position): string | undefined {
        return this.map.index(position)
    }

    public isInside(position: Position): boolean {
        return this.map.inside(position)
    }

    public getTileByIndex(index: string): Tile | undefined {
        return this.map.getByIndex(index)
    }

    public getDrawablesByIndex(index: string): Drawable[] {
        return this.drawables.get(index) || []
    }

    public buildDoor(world: World, position: Position, room: number): void {
        this.setTile(world, position, corridorTile(room))
        world.entity()
            .with("position", new Boxed(position))
            .with("drawable", new Drawable("+", primary[2]))
            .with("description", new Boxed("a door"))
            .with("lightBlocking")
            .with("blocking")
            .with("trigger", doorTrigger)
            .close()
    }

    public buildRoom(world: World, rectangle: Rectangle): number {
        const room = world.entity()
            .with("room", { shape: rectangle })
            .close()
        foreach(rasterizeRectangle(rectangle, true), (p: Vector2D) =>
            this.setTile(world, new Position(this.activeDomain!, p), roomTile(room))
        )
        this.addLightToRoom(world, room, rectangle)
        return room
    }

    public buildHub(world: World, rectangle: Rectangle): number {
        const room = world.entity()
            .with("room", { shape: rectangle })
            .close()
        foreach(rasterizeRectangle(rectangle, true), (p: Vector2D) =>
            this.setTile(world, new Position(this.activeDomain!, p), hubTile(room))
        )
        this.addLightToRoom(world, room, rectangle)
        return room
    }

    public addLightToRoom(world: World, room: number, rectangle: Rectangle): void {
        const lighting: LightingSystem | undefined = world.systems.get(LightingSystem.NAME) as LightingSystem | undefined
        if (lighting !== undefined) {
            const light = lighting.buildLight(world, new Position(this.activeDomain!, rectangle.mid))
            world.entity(room).rel(r => r.with("contains").to(light).close()).close()
        }
    }

    public openCorridor(world: World): number {
        return world.entity().close()
    }

    public closeCorridor(world: World, room: number, rectangle: Rectangle): number {
        world.entity(room)
            .with("room", { shape: rectangle })
            .close()

        this.addLightToRoom(world, room, rectangle)
        return room
    }

    public buildAsset(world: World, position: Position, asset: (Tile | undefined)[]): void {
        for (let y = 0; y < 4; y++) {
            for (let x = 0; x < 4; x++) {
                const tile: Tile | undefined = asset[4 * y + x]
                if (tile !== undefined) {
                    this.setTile(world, new Position(this.activeDomain!, new Vector2D(position.x + x - 1, position.y + y - 1)), tile)
                }
            }
        }
    }

    public isFree(positions: Position[], roomFilter?: number): boolean {
        return positions.find(p => !this.isInside(p) || this.isRoom(p, roomFilter)) === undefined
    }

    public isWalls(positions: Position[]): boolean {
        return positions.find(p => !this.isInside(p) || !this.isWall(p)) === undefined
    }

    public isRoom(p: Position, roomFilter?: number): boolean {
        const tile: Tile | undefined = this.getTile(p)
        return tile !== undefined && tile.room !== undefined && tile.room !== roomFilter
    }

    public isWall(p: Position): boolean {
        const tile: Tile | undefined = this.getTile(p)
        return tile !== undefined && tile.character === "#"
    }

    public isBlocking(world: World, position: Position, entityFilter: number): boolean {
        const tile: Tile | undefined = this.getTile(position)
        const tileBlocking = tile === undefined || tile.blocking
        if (tileBlocking) {
            return true
        }
        const entityBlocking = !world.fetch()
            .on(t => t
                .ofPartition("position", new Boxed(position))
                .hasLabel("blocking")
            )
            .stream()
            .filter(e => e.entity !== entityFilter)
            .isEmpty()
        if (entityBlocking) {
            return true
        }
        return false
    }

    private fillWalls(domain: Domain): void {
        const neighbours = toArray(rasterizeRectangle(Rectangle.from(new Vector2D(-1, -1), new Size(2, 2)), true))
        const newWallPositions: Vector2D[] = []
        foreach(rasterizeRectangle(Rectangle.from(new Vector2D(0, 0), this.boundary), true), (p: Vector2D) => {
            if (this.map.get(new Position(domain, p)) === undefined) {
                const firstDefinedTile = neighbours.find(offset =>
                    this.map.get(new Position(domain, p.add(offset))) !== undefined
                )
                if (firstDefinedTile !== undefined) {
                    newWallPositions.push(p)
                }
            }
        })
        newWallPositions.forEach(p => this.map.set(new Position(domain, p), wallTile()))
    }

    private drawDrawables(world: World, display: Display, viewport: Viewport): void {
        const lighting: LightingSystem | undefined = world.systems.get(LightingSystem.NAME) as LightingSystem | undefined
        if (lighting === undefined) {
            return
        }

        const topLeft = viewport.topLeft
        const allDrawables: DrawableWithData[] = world
            .fetch()
            .on(v => v.hasLabel("drawable").hasLabel("position")
                .matchesValue("position", (p: Boxed<Position>) => {
                    const index: string | undefined = this.getIndex(p.value)
                    if (index !== undefined && lighting.isDiscovered(index)) {
                        return viewport.rectangle.isInside(p.value.toVector2D())
                    } else {
                        return false
                    }
                })
            )
            .withComponents("position", "drawable", "description")
            .collect()

        const alreadyDrawn = new Set<string>()
        allDrawables.forEach((comp: DrawableWithData) => {
            const p = comp.position.value.subtract(topLeft)
            const index: string = this.getIndex(p)!
            const color = lighting.getColor(comp.drawable)
            this.renderCharacter(display, p, comp.drawable.character, color)
            alreadyDrawn.add(index)
        })

        const menu: MenuSystem | undefined = world.systems.get(MenuSystem.NAME) as MenuSystem | undefined
        if (menu !== undefined && menu.activeMenuItem === MenuItems.Map) {
            this.drawDescriptions(allDrawables, alreadyDrawn, topLeft, display)
        }
    }

    private drawDescriptions(drawables: DrawableWithData[], alreadyDrawn: Set<string>, topLeft: Vector2D, display: Display): void {
        drawables.forEach((comp: DrawableWithData) => {
            if (comp.description !== undefined) {
                this.tryToDrawDescription(display,
                    comp.position.value.toVector2D().subtract(topLeft),
                    comp.description.value,
                    alreadyDrawn
                )
            }
        })
    }

    private renderTile(display: Display, position: Position, tile: Tile, color: Color): void {
        if (tile.room !== undefined && tile.room === this.selectedRoom) {
            this.renderCharacter(display, position, tile.character, color, gray[3].rgb)
        } else {
            this.renderCharacter(display, position, tile.character, color)
        }
    }

    private renderCharacter(display: Display, position: Position, character: string, color: Color, bg?: string): void {
        display.draw(position.x, position.y, character, color.rgb, bg)
    }

    private tryToDrawDescription(display: Display, position: Vector2D, description: string, alreadyDrawn: Set<string>): void {
        if (this.canDraw(position, description, alreadyDrawn)) {
            display.draw(position.x + 1, position.y, gridSymbols[0])
            for (let offset = 0; offset < description.length; offset++) {
                const p = new Position(this.activeDomain!, new Vector2D(position.x + 2 + offset, position.y))
                const index = this.getIndex(p)
                if (index !== undefined) {
                    alreadyDrawn.add(index)
                    display.draw(p.x, p.y, description[offset], gray[0].rgb, gray[2].rgb)
                }
            }
        }
    }

    private canDraw(position: Vector2D, text: string, alreadyDrawn: Set<string>): boolean {
        for (let offset = 0; offset < text.length; offset++) {
            const p = new Position(this.activeDomain!, new Vector2D(position.x + 2 + offset, position.y))
            const index = this.getIndex(p)
            if (index === undefined || alreadyDrawn.has(index)) {
                return false
            }
        }
        return true
    }
}
