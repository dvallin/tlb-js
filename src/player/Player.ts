import { GameSettings } from "@/Game"
import { Position, Domain } from "@/geometry/Position"
import { GameSystem, RenderLayer } from "@/systems/GameSystem"
import { World, MapStorage, Boxed } from "mogwai-ecs/lib"
import { Display } from "rot-js"

import { Tile } from "@/map/Tile"
import { Drawable } from "@/rendering/Drawable"
import { Input } from "@/systems/Input"
import { MenuSystem, MenuItems } from "@/menu/Menu"
import { MapSystem } from "@/map/Map"
import { Color } from "@/rendering/Color"
import { Vector2D } from "@/geometry/Vector2D"

export class PlayerSystem implements GameSystem {

    public static NAME: string = "player"

    public static fromSettings(settings: GameSettings): PlayerSystem {
        return new PlayerSystem(settings.map_width)
    }

    public renderLayer: RenderLayer = RenderLayer.None

    public readonly playerSpawn: Position

    private constructor(width: number) {
        this.playerSpawn = new Position(Domain.Tower, new Vector2D(Math.floor(width / 2), 0))
    }

    public register(world: World): void {
        world.registerSystem(PlayerSystem.NAME, this)
        world.registerComponent("player")
        world.registerComponent("blocking")
        world.registerComponent("active", new MapStorage<Tile>())
        world.registerComponent("drawable", new MapStorage<Drawable>())
    }

    public build(world: World): void {
        world.entity()
            .with("player")
            .with("position", new Boxed<Position>(this.playerSpawn))
            .with("blocking")
            .with("drawable", new Drawable("@", new Color([255, 255, 255])))
            .with("active")
            .close()
    }

    public execute(world: World): void {
        const input: Input | undefined = world.systems.get(Input.NAME) as Input | undefined
        const menu: MenuSystem | undefined = world.systems.get(MenuSystem.NAME) as MenuSystem | undefined
        const map: MapSystem | undefined = world.systems.get(MapSystem.NAME) as MapSystem | undefined
        if (input !== undefined && menu !== undefined && map !== undefined) {
            if (menu.activeMenuItem === MenuItems.Player) {
                const delta = input.movementDelta()
                if (delta.x !== 0 || delta.y !== 0) {
                    const player: { entity: number, position: Boxed<Position> } | undefined = world
                        .fetch()
                        .on(t => t.hasLabel("player").hasLabel("active"))
                        .withComponents("position")
                        .first()
                    if (player !== undefined) {
                        const scaledDelta = delta.normalize().mult(0.4).fround()
                        const newPosition = player.position.value.add(scaledDelta)
                        if (map.isInside(newPosition) && !map.isBlocking(world, newPosition, player.entity)) {
                            player.position.value = newPosition
                        }
                    }
                }
            }
        }
    }

    public render({ }: World, { }: Display): void {
        //
    }

    public afterRender({ }: World): void {
        //
    }
}
