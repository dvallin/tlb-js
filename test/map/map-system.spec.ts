import { MapSystem } from "../../src/map/Map"
import { PlayerSystem } from "../../src/player/Player"
import { LightingSystem } from "../../src/lighting/Lighting"
import { ViewportSystem } from "../../src/rendering/Viewport"
import { World } from "mogwai-ecs"
import { Game } from "../../src/Game"
import { lightDisplay, tileDisplay } from "../testing/display-mocks"
import { Display } from "rot-js"

describe("MapSystem", () => {

    it("renders tiles", () => {
        const img = []

        renderOneTick(tileDisplay(img), true)

        expect(render(img, 60, 30)).toMatchSnapshot()
    })

    it("renders only discovered tiles", () => {
        const img = []

        renderOneTick(tileDisplay(img), false)

        expect(render(img, 60, 30)).toMatchSnapshot()
    })

    it("renders lighting", () => {
        const light = []

        renderOneTick(lightDisplay(light), true)

        expect(render(light, 60, 30)).toMatchSnapshot()
    })
})

function renderOneTick(display: Display, allVisible: boolean): void {
    const world = new World()
    const game = new Game({
        screen_width: 60, screen_height: 30,
        map_width: 60, map_height: 30,
        framerate: 30,
        seed: "default-seed"
    }, display, world)
    game.addGameSystem(ViewportSystem.fromSettings(game.settings))
    game.addGameSystem(PlayerSystem.fromSettings(game.settings))
    const lighting = LightingSystem.fromSettings(game.settings)
    if (allVisible) {
        lighting.isDiscovered = () => (true)
        lighting.isVisible = () => (true)
    }
    game.addGameSystem(lighting)
    game.addGameSystem(MapSystem.fromSettings(game.settings))

    game.build()
    game.tick()
}

function render(img: string[][], width: number, height: number): string {
    let result = ""
    for (let y = 0; y < height; ++y) {
        const line = img[y] || []
        for (let x = 0; x < width; ++x) {
            result += line[x] || " "
        }
        result += "\n"
    }
    return result
}
