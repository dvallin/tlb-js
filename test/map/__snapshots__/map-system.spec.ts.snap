// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapSystem renders lighting 1`] = `
"                                                            
                                                            
                                                            
                                                            
                                                            
                                                            
                            18981                           
                            188811111111111111111111111111  
                  1122222221299984444444444444444444444441  
                  1899999998299921111111111111111111111141  
                  189999999829992                      141  
                  189992999899992                111111141  
                  189999999839992        12222222144444441  
                  189999999839992        19999999141111111  
                  112222222129992        19999999141        
       1111222211         112999211      19992999841        
       1666778872222222222278999761      19999999141        
       1677788889999999999989999861      19999999142222222  
       1677228889999999999989922861      12222222142999992  
       1672992889999999999989299271              142999992  
       1672992889999999999988299271              149992992  
       1677228872233243222277722761   11111111111142999992  
       1666776661188888881166666661   18888888881142999992  
       1111811111199999991166666661   18999999981142222222  
    111111181    19999999112222222211118999999981141        
    299999271    199999991 189999999811899999998114111111111
    299999261    199929991 189999999811899929998114188888881
    299999251    199999991 189999999811899999998114199999991
    299299941    199999991 189992999811899999998114199999991
    299999241    199999991 189999999811899999998114199999991
"
`;

exports[`MapSystem renders only discovered tiles 1`] = `
"                                                            
                                                            
                                                            
                                                            
                                                            
                                                            
                            #.@.#                           
                            #...#                           
                            #.....                          
                            #...###                         
                            #...#                           
                            +...#                           
                            #...#                           
                            #...#                           
                            #...#                           
                            #...#                           
                            .....                           
                            .....                           
                            ..╓–.                           
                            .│██╕                           
                            .╘██│                           
                            ..–╜.                           
                           .......                          
                           .......                          
                           #######                          
                                                            
                                                            
                                                            
                                                            
                                                            
"
`;

exports[`MapSystem renders tiles 1`] = `
"                                                            
                                                            
                                                            
                                                            
                                                            
                                                            
                            #.@.#                           
                            #...##########################  
                  ###########............................#  
                  #.........#...########################.#  
                  #.........#...#                      #.#  
                  #....░....+...#                #######.#  
                  #.........#...#        #########.......#  
                  #.........#...#        #.......#.#######  
                  ###########...#        #.......#.#        
       ##########         ###...###      #...¬...+.#        
       #........###########.......#      #.......#.#        
       #..........................#      #.......#.#######  
       #...╓–.................╓–..#      #########.#.....#  
       #..│██╕...............│██╕.#              #.#.....#  
       #..╘██│...............╘██│.#              #.+..Ô..#  
       #...–╜...#####+#####...–╜..#   ############.#.....#  
       #........##.......##.......#   #.........##.#.....#  
       ####.######.......##.......#   #.........##.#######  
    #######.#    #.......##############.........##.#        
    #.....#.#    #.......# #.........##.........##.#########
    #.....#.#    #...Ó...# #.........##....∞....##.#.......#
    #.....#.#    #.......# #.........##.........##.#.......#
    #..‼..+.#    #.......# #....…....##.........##.#.......#
    #.....#.#    #.......# #.........##.........##.#.......#
"
`;
